### PR TITLE
Introduce useAssistant (experimental).

### PR DESCRIPTION
**Pull request by lgrammel**
## Setup

Follow instructions in `examples/next-openai/app/api/assistant/assistant-setup.md`

## Example

https://github.com/vercel/ai/assets/205036/dc773902-139d-4e2d-9cc2-ee7269d75342

### CODE REVIEW

File: examples/next-openai/app/api/assistant/AssistantResponse.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      const textEncoder = new TextEncoder();
+
+      await process({
+        // TODO send custom data
+
+        sendMessage: (message: AssistantMessage) => {
+          // TODO have a smarter streaming protocol that only sends delta + msg id
+          controller.enqueue(
+            textEncoder.encode(`0: ${JSON.stringify(message)}\n\n`),
+          );
```
Comment (MaxLeiter) [inline examples/next-openai/app/api/assistant/AssistantResponse.ts:35]: nit: we should use the same utils we use in stream data

Comment (lgrammel) [reply examples/next-openai/app/api/assistant/AssistantResponse.ts:42]: 👍🏻

**Comment by oayandosu** [comment]
Can't wait for this to land!

**Comment by tgonzales** [comment]
It seems reckless to me to put an infinite While in api layer, an assistant could receive several tools. One approach could be to put the while in useAssistant and send the response via callback, for example. useAssistant({api: '/api/assistant', callback:'/api/tools'})

File: packages/core/streams/assistant-response.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { AssistantMessage } from '../shared/types';
+
+export function expertimental_AssistantResponse(
```
Comment (burn2delete) [inline packages/core/streams/assistant-response.ts:3]: typo `expertimental`

File: packages/core/react/use-assistant.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
+import { nanoid } from 'nanoid';
+import { useState } from 'react';
+import { processMessageStream } from '../shared/process-message-stream';
+import { Message } from '../shared/types';
+
+export type AssistantStatus = 'in_progress' | 'awaiting_message';
+
+export function useAssistant_experimental({
```
Comment (burn2delete) [inline packages/core/react/use-assistant.ts:8]: should this be `experimental_useAssistant`?

File: examples/next-openai/app/api/assistant/route.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  // Parse the request body
+  const input: {
+    threadId: string | null;
+    message: string;
+  } = await req.json();
+
+  let threadId = input.threadId;
+  const message = input.message;
+
+  return expertimental_AssistantResponse(
```
Comment (burn2delete) [inline examples/next-openai/app/api/assistant/route.ts:31]: typo

**Comment by burn2delete** [comment]
I would remove all instances of `let` and replace the logic with something more functional. I think the while loops should probably be dropped as well.

File: examples/next-openai/app/api/assistant/assistant-setup.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        "type": "string",
+        "enum": ["bedroom", "home office", "living room", "kitchen", "bathroom"]
+      }
+    },
+    "required": ["location"]
+  }
+}
+```
+
+### setRooomTemperature function
```
Comment (snettah) [inline examples/next-openai/app/api/assistant/assistant-setup.md:40]: typo setRoomTemperature

File: packages/core/react/use-assistant.ts (hunk 3 above)
Comment (lgrammel) [reply packages/core/react/use-assistant.ts:8]: That's what I used initially - it's prevented by the hooks linting rule.

**Comment by lgrammel** [comment]
> It seems reckless to me to put an infinite While in api layer, an assistant could receive several tools. One approach could be to put the while in useAssistant and send the response via callback, for example. useAssistant({api: '/api/assistant', callback:'/api/tools'})

`useAssistant` is on the client side. The goal is that all assistant based-behavior can run on the server side, without any client involvement. The client is only notified with information that could lead to rendering changes (e.g. progress information, streaming messages in the future, etc.)

Can you point to the exact while loop? Please note that a) the server side implementation is an example, any production system would use something way more complex to guarantee resiliency, failover, etc., and b) most of the code has async calls.

File: packages/core/react/use-assistant.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    });
+
+    if (result.body == null) {
+      throw new Error('The response body is empty.');
+    }
+
+    await processMessageStream(result.body.getReader(), (message: string) => {
+      try {
+        const { type, value } = getStreamStringTypeAndValue(message);
+        const messageContent = value as any;
```
Comment (lgrammel) [inline packages/core/react/use-assistant.ts:58]: Increasing type safety here would be nice. Ideally this would happen in `getStreamStringTypeAndValue`

**Comment by teambraveshine** [comment]
Very excited about this y'all. Thanks for putting in the hard work.

**Comment by mannybecerra** [comment]
Awesome work @lgrammel ✨

**Comment by Edjevw12** [comment]
Looks really good!

File: examples/next-openai/app/api/assistant/assistant-setup.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  "description": "Get the temperature in a room",
+  "parameters": {
+    "type": "object",
+    "properties": {
+      "room": {
+        "type": "string",
+        "enum": ["bedroom", "home office", "living room", "kitchen", "bathroom"]
+      }
+    },
+    "required": ["location"]
```
Comment (snettah) [inline examples/next-openai/app/api/assistant/assistant-setup.md:35]:
```suggestion
    "required": ["room"]
```

File: examples/next-openai/app/api/assistant/assistant-setup.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  "parameters": {
+    "type": "object",
+    "properties": {
+      "room": {
+        "type": "string",
+        "enum": ["bedroom", "home office", "living room", "kitchen", "bathroom"]
+      },
+      "temperature": { "type": "number" }
+    },
+    "required": ["location", "temperature"]
```
Comment (snettah) [inline examples/next-openai/app/api/assistant/assistant-setup.md:55]:
```suggestion
    "required": ["room", "temperature"]
```

**Comment by DystopianDisco** [comment]
great work @lgrammel.

"handeFileUpload" massive.

looking forward to this merging soon.

File: packages/core/shared/stream-parts.ts [hunk 9]
```diff
@@ -90,12 +174,47 @@ type StreamPartValueType = {
 export type StreamPartType =
   | ReturnType<typeof textStreamPart.parse>
   | ReturnType<typeof functionCallStreamPart.parse>
-  | ReturnType<typeof dataStreamPart.parse>;
+  | ReturnType<typeof dataStreamPart.parse>
+  | ReturnType<typeof errorStreamPart.parse>
+  | ReturnType<typeof assistantMessage.parse>
+  | ReturnType<typeof assistantControlData.parse>;
```
Comment (lgrammel) [inline packages/core/shared/stream-parts.ts:180]: I'll look into reducing this code duplication (also in other places) in future PRs.

File: packages/core/shared/stream-parts.ts [hunk 10]
```diff
@@ -90,12 +174,47 @@ type StreamPartValueType = {
...
+ * 1: {"function_call": {"name": "get_current_weather", "arguments": "{\\n\\"location\\": \\"Charlottesville, Virginia\\",\\n\\"format\\": \\"celsius\\"\\n}"}}
+ *```
+ */
+export const StreamStringPrefixes = {
+  [textStreamPart.name]: textStreamPart.code,
+  [functionCallStreamPart.name]: functionCallStreamPart.code,
+  [dataStreamPart.name]: dataStreamPart.code,
+  [errorStreamPart.name]: errorStreamPart.code,
+  [assistantMessage.name]: assistantMessage.code,
+  [assistantControlData.name]: assistantControlData.code,
```
Comment (lgrammel) [inline packages/core/shared/stream-parts.ts:217]: Move from `utils.ts` to have all duplicated code in a single file for now.

File: packages/core/shared/stream-parts.ts [hunk 11]
```diff
@@ -108,21 +227,21 @@ export const validCodes = streamParts.map(part => part.code);
  * @throws An error if the string cannot be parsed.
  */
 export const parseStreamPart = (line: string): StreamPartType => {
-  const firstSeperatorIndex = line.indexOf(':');
+  const firstSeparatorIndex = line.indexOf(':');
```
Comment (lgrammel) [inline packages/core/shared/stream-parts.ts:230]: Fixed typo.

File: packages/core/react/use-assistant.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable react-hooks/rules-of-hooks */
```
Comment (lgrammel) [inline packages/core/react/use-assistant.ts:1]: Required to use the name `experimental_*` for a hook.

File: examples/next-openai/app/api/assistant/route.ts [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+  return experimental_AssistantResponse(
+    { threadId, messageId: createdMessage.id },
+    async ({ threadId, sendMessage }) => {
+      // Run the assistant on the thread
+      const run = await openai.beta.threads.runs.create(threadId, {
+        assistant_id:
+          process.env.ASSISTANT_ID ??
+          (() => {
+            throw new Error('ASSISTANT_ID is not set');
```
[the remark this answers is not in the document]
Comment (SharadKumar) [reply examples/next-openai/app/api/assistant/route.ts:45]: In fact working on a client project where there are multple explicit OpenAI threads, each having its own assistant/files etc.

**Review by SharadKumar — CHANGES_REQUESTED** [review]

File: examples/next-openai/app/api/assistant/route.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (lgrammel) [reply examples/next-openai/app/api/assistant/route.ts:45]: it's an example

**Comment by codinginflow** [comment]
Can you expose the `threadId` from the hook so we can store a list of threads client-side?

**Comment by lgrammel** [comment]
> Can you expose the `threadId` from the hook so we can store a list of threads client-side?

@codinginflow https://github.com/vercel/ai/pull/755

**Comment by ibrahimyaacob92** [comment]
when I pass thread id in the useAssistant hook, I was expecting that I would get the previous messages, but i don't

**Pull request merged**